src/Microsoft.ML.StandardLearners/FactorizationMachine/FactorizationMachine.cs

```diff
@@ -0,0 +1,711 @@
+using Microsoft.ML.Runtime.Training;
+
+[assembly: LoadableClass(FieldAwareFactorizationMachineTrainer.Summary, typeof(FieldAwareFactorizationMachineTrainer), typeof(FieldAwareFactorizationMachineTrainer.Arguments),
+    new[] { typeof(SignatureBinaryClassifierTrainer), typeof(SignatureTrainer) }, FieldAwareFactorizationMachineTrainer.UserName, FieldAwareFactorizationMachineTrainer.LoadName,
+    FieldAwareFactorizationMachineTrainer.ShortName, DocName = "trainer/FactorizationMachine.md")]
+
+[assembly: LoadableClass(typeof(FieldAwareFactorizationMachinePredictor), null, typeof(SignatureLoadModel), "Field Aware Factorization Machine", FieldAwareFactorizationMachinePredictor.LoaderSignature)]
+
+[assembly: LoadableClass(typeof(void), typeof(FieldAwareFactorizationMachineTrainer), null, typeof(SignatureEntryPointModule), FieldAwareFactorizationMachineTrainer.LoadName)]
+
+namespace Microsoft.ML.Runtime.FactorizationMachine
+{
+    internal sealed class FieldAwareFactorizationMachineUtils
+    {
+        internal static int GetAlignedVectorLength(int length)
+        {
+            int res = length % 4;
+            if (res == 0)
+                return length;
+            else
+                return length + (4 - res);
+        }
+
+        internal static bool LoadOneExampleIntoBuffer(ValueGetter<VBuffer<float>>[] getters, VBuffer<float> featureBuffer, bool norm, ref int count,
+            int[] fieldIndexBuffer, int[] featureIndexBuffer, float[] featureValueBuffer)
+        {
+            count = 0;
+            float featureNorm = 0;
+            int bias = 0;
+            float annihilation = 0;
+            for (int f = 0; f < getters.Length; f++)
+            {
+                getters[f](ref featureBuffer);
+                foreach (var pair in featureBuffer.Items())
+                {
+                    fieldIndexBuffer[count] = f;
+                    featureIndexBuffer[count] = bias + pair.Key;
+                    featureValueBuffer[count] = pair.Value;
+                    featureNorm += pair.Value * pair.Value;
+                    annihilation += pair.Value - pair.Value;
+                    count++;
+                }
+                bias += featureBuffer.Length;
+            }
+            featureNorm = MathUtils.Sqrt(featureNorm);
+            if (norm)
+            {
+                for (int i = 0; i < count; i++)
+                    featureValueBuffer[i] /= featureNorm;
+            }
+            return FloatUtils.IsFinite(annihilation);
+        }
+    }
+
+    /// <summary>
+    /// Train a field-aware factorization machine using ADAGRAD (an advanced stochastic gradient method). See references below
+    /// for details. This trainer is essentially faster the one introduced in [2] because of some implemtation tricks[3].
+    /// [1] http://jmlr.org/papers/volume12/duchi11a/duchi11a.pdf
+    /// [2] http://www.csie.ntu.edu.tw/~cjlin/papers/ffm.pdf
+    /// [3] fast-ffm.tex in FactorizationMachine project folder
```

Comment:
Please cite https://github.com/wschin/fast-ffm/blob/master/fast-ffm.pdf for [3]. Thanks.